src/corelibs/U2Core/src/gobjects/MultipleChromatogramAlignmentObject.cpp

```diff
@@ -260,6 +260,47 @@ void MultipleChromatogramAlignmentObject::trimRow(const int rowIndex, int curren
     updateCachedMultipleAlignment(modificationInfo);
 }
 
+void MultipleChromatogramAlignmentObject::updateAlternativeMutations(const McaReadsTabSettings& settings, U2OpStatus& os) {
+    for (int i = 0; i < getNumRows(); i++) {
+        const MultipleChromatogramAlignmentRow& mcaRow = static_cast<const MultipleChromatogramAlignmentRow&>(getRow(i));
+        const int ungappedLength = mcaRow->getUngappedLength();
+
+        QMap<qint64, char> newCharList;
+        for (int j = 0; j < ungappedLength; j++) {
+            auto res = mcaRow->getTwoHighestPeaks(j);
+            const double minimumThresholdValue = (double)res.second.value / res.first.value * 100;
+            char newChar = 'A';
+            if (minimumThresholdValue < settings.threshold || !settings.showAlternativeMutations) {
+                newChar = DNAChromatogram::TRACE_CHARACTER[res.first.trace];
+            } else {
+                newChar = DNAChromatogram::TRACE_CHARACTER[res.second.trace];
+            }
+
+            int gappedPos = mcaRow->getGappedPosition(j);
+            char currentChar = mcaRow->charAt(gappedPos);
+            if (currentChar == newChar) {
+                continue;
+            }
+
+            newCharList.insert(gappedPos, newChar);
+        }
+
+        const MultipleAlignment& ma = getMultipleAlignment();
+        qint64 modifiedRowId = ma->getRow(i)->getRowId();
+        McaDbiUtils::replaceCharactersInRow(getEntityRef(), modifiedRowId, newCharList, os);
+        SAFE_POINT_OP(os, );
+    }
+    updateCachedMultipleAlignment();
+}
+
+void MultipleChromatogramAlignmentObject::sl_viewIsAbout2BeDestroyed() {
```

Comment:
What if I have 2 views opened with different settings?

Reply:
`MultipleChromatogramAlignmentObject` instance "exists" in one view. If you open two Sanger Reads Editor, you will have two instances of `MultipleChromatogramAlignmentObject`, each of them have their own settings saved.

Reply:
This looks like a big issue: we mix persistent & view data

All GObjects in the code are data objects. A view is allowed to modify these objects, but in this case this modification is persistent (should not be rolled back on view close) and should be visible to all other views.

Building a new functionality on incorrectly written base will produce even more tech debt than before.

Here is some questions and ideas: 
1) do we store MCA object in the database when it is added to a project and no view is opened?
2) If 1) is yes - does the MCA view clones this object before use?
3) if 2) is yes - the view should create its own in-memory DBI with a cloned object and does not need to 'rollback' anything on close.

Reply:
1. Chromatogram files are databases itself (they are stored in the "ugenedb" format, which is SQL database), so MultipleChromatogramAlignmentObject data are stored in there permanently. 
2. Sorry, don' get this question. Could you detail it?

Reply:
OK, the main concern/question is why do you want to do anything when the view is closed?

Reply:
**Scenario**:
1. Open ".ugenedb" in Sanger Reads Editor.
2. Open the "Reads" tab, turn "Alternative mutations" on and decrease the "Threshold" value.
    Expected: some characters has been replaced with alternative.
3. Close **the project** (or just remove the ".ugenedb" file from the project).
4. Open the same ".ugenedb" file in Sanger Reads Editor again.
    Expected: **no** alternative mutations are shown, the "Alternative mutations" feature on the "Reads" tab **is disabled**.

`MultipleChromatogramAlignmentObject::sl_viewIsAbout2BeDestroyed` is caught when we remove file from the project. It makes alternative mutations not visible. If we don't have this slot, we will have the following state.
    State: alternative mutations from point "2." are **still visible**, but the "Alternative mutations" feature on the "Reads" tab **is disabled**.

Reply:
When we remove a document from the project a signal is emitted that the document is removed from the project. 
Why don't you use this signal instead?

Reply:
What signal exactly? I couldn't find something fit

Reply:
> Close the project (or just remove the ".ugenedb" file from the project).

in ProjectModel.h
```
signals:
    void si_documentAdded(Document *d);
    void si_documentRemoved(Document *d);
```

Reply:
It works when you remove a doc from the project, but if you simply close the project it doesn't work

Reply:
I understand about the view, but I quoted your text and answered your question.
The big problem with your patch that you have not separated what is View and what is Data.

If the view changes the state (turns on alternative bases) it has 2 options:

Option 1) Do not change the data: keep the change "view-local"
In this case you do not need the "on-close-callback"

Option 2) The view can change the data (like rename row) but in this case the change to the data is persistent.
In this case you do not need the "on-close-callback" again.

Looks like you implemented some other non-standard option that will cause a lot of issues.

Here is another issue:
1) Open view A
2) Open view B of the same MCA
3) Change the view mode to show alternative mutations in A
4) Switch to B: I see the alternative mutations too, but the options panel (view settings) does not show the correct state.

Check the screenshots (options panel, base at 298 and active view tab):
![mca-1](https://user-images.githubusercontent.com/17791039/117915873-1ddfdd80-b2b4-11eb-8be6-b982e5446181.png)

![mca-2](https://user-images.githubusercontent.com/17791039/117915874-1ddfdd80-b2b4-11eb-9a4d-70a4302791d8.png)



I would suggest you to implement either Option-1 or Option-2: make the change view-local or persistent.